digitalearthau/coherence.py

```diff
@@ -139,18 +194,80 @@ def _check_ancestors(check_ancestors: bool,
                     if archive_siblings:
-                        ancestors_archive_count += _archive_duplicate_siblings(dc, sibling_ids + [str(dataset.id)])
+                        _archive_count += _archive_duplicate_siblings(dc, sibling_ids + [str(dataset.id)])
+
+
+def _manage_downstream_ds(dc: Datacube,
+                          dataset: Dataset):
+    """
+    Need to manage the following two scenarios, with locationless source datasets:
+       1) Identify and list all the downstream datasets (may include level2, NBAR, PQ, Albers, WOfS, FC) that are either
+          Active or Archived.
+       2) Identify and list all the downstream datasets (includes summary products) that are either Active or Archived.
+    """
+
+    # Fetch all the derived datasets using the source dataset.
+    derived_dataset = _fetch_derived_datasets(dataset, dc)
+
+    for dataset_list in derived_dataset:
+        for d in dataset_list:
+            if d.uris is None:
+                # Append locationless source dataset to the list
+                _bad_downstream_dataset.append(dataset)
+
+                # Fetch all downstream datasets and append to the list
+                _process_bad_derived_datasets(dataset, d, _fetch_derived_datasets(d, dc))
+
+
+def _process_bad_derived_datasets(source_ds, dataset, derived_dataset):
+    global _downstream_dataset_count, _downstream_summary_prod_count
+
+    _LOG.info("locationless." + str(source_ds.type.name) + ".dry_run (Dataset ID: %s)" % str(source_ds.id),
+              downstream_dataset_id=str(dataset.id),
+              downstream_dataset_type=str(dataset.type.name),
+              downstream_dataset_location=str(dataset.uris))
+    _downstream_dataset_count += 1
+
+    for dataset_list in derived_dataset:
+        for d_datasets in dataset_list:
+            # Exclude derived products such as summary products, FC_percentile products, etc.
+            if d_datasets.type.name in PRODUCT_TYPE_LIST:
+                _LOG.info("downstream." + str(dataset.type.name) + ".dry_run (Dataset ID: %s)" % str(dataset.id),
+                          downstream_dataset_id=str(d_datasets.id),
+                          downstream_dataset_type=str(d_datasets.type.name),
+                          downstream_dataset_location=str(d_datasets.uris))
+                _downstream_dataset_count += 1
+            else:
+                _LOG.info("derived." + str(dataset.type.name) + ".dry_run (Dataset ID: %s)" % str(dataset.id),
+                          downstream_dataset_id=str(d_datasets.id),
+                          dataset_type=str(d_datasets.type.name))
+                _downstream_summary_prod_count += 1
+
+            # Append all the downstream datasets derived from locationless source dataset to the list
+            _bad_downstream_dataset.append(d_datasets)
+
+
+def _fetch_derived_datasets(dataset, dc):
+    to_process = {dataset.id}
+    derived_dataset = collections.deque({})
```

Comment:
This would be better as `derived_datasets = set()`, and then using `derived_datasets.update(derived)` below will give you a single set of all the derived datasets, instead of the current collection of collections, leading to the nested `for` loops on line 223-4 and 242-3.